### ISSUE TITLE
Console.SetWindowPosition(int left, int top) throws IOException complaining about incorrect parameter from within the Win32 API

### ISSUE THREAD

**Issue by daxian-dbw** [dotnet/runtime#31949]
`Console.SetWindowPosition(int left, int top)` throws `System.IO.IOException: The parameter is incorrect.` from within the Win32 API.
This was reported by [a dozen of issues](https://github.com/PowerShell/PSReadLine/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+SetWindowPosition) in PowerShell/PSReadLine repo. Here is one of them: https://github.com/PowerShell/PSReadLine/issues/551, and there are more recent ones like https://github.com/PowerShell/PSReadLine/issues/1328 and https://github.com/PowerShell/PSReadLine/issues/1341

We reviewed the code in PSReadLine and .NET Core, and this seems like an impossible condition, quoted the comment from https://github.com/PowerShell/PSReadLine/issues/551#issuecomment-341856386:
> The .Net api Console.SetWindowsPosition validates the parameters before calling the Win32 api, yet it's the Win32 api reporting the error.

`Console.SetWindowPosition(int left, int top)` calls to `ConsolePal.SetWindowPosition(left, top)`, which does the [argument validations](https://github.com/dotnet/runtime/blob/0b25063ec056b0d6668bf0848d108d8c873f3fce/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L941-L947) before calling into the Win32 API `Interop.Kernel32.SetConsoleWindowInfo`:

            // Check for arithmetic underflows & overflows.
            int newRight = left + srWindow.Right - srWindow.Left + 1;
            if (left < 0 || newRight > csbi.dwSize.X || newRight < 0)
                throw new ArgumentOutOfRangeException(nameof(left), left, SR.ArgumentOutOfRange_ConsoleWindowPos);
            int newBottom = top + srWindow.Bottom - srWindow.Top + 1;
            if (top < 0 || newBottom > csbi.dwSize.Y || newBottom < 0)
                throw new ArgumentOutOfRangeException(nameof(top), top, SR.ArgumentOutOfRange_ConsoleWindowPos);

-------------------------------

I don't have a way to reproduce this, but the reported issues in PSReadLine repository suggest that it is not that rare to happen. So I choose to report it here and hope to get some guidance.

**Comment by danmosemsft** [comment]
I can reproduce. I think there is an off by 1 somewhere - it's pretty tricky

**Comment by daxian-dbw** [comment]
Thanks @danmosemsft for the prompt response! Can you please share the repro to me?

**Comment by danmosemsft** [comment]
Sure, I just made a C# console app, with `Console.SetWindowPosition(0, 0);`, put a breakpoint on that and dragged the bottom of the console window to the top so height is zero, then continued.

**Comment by danmosemsft** [comment]
[Windows says](https://docs.microsoft.com/en-us/windows/console/setconsolewindowinfo)

>The function fails if the specified window rectangle extends beyond the boundaries of the console screen buffer. This means that the Top and Left members of the lpConsoleWindow rectangle (or the calculated top and left coordinates, if bAbsolute is FALSE) cannot be less than zero. Similarly, the Bottom and Right members (or the calculated bottom and right coordinates) cannot be greater than (screen buffer height – 1) and (screen buffer width – 1), respectively. The function also fails if the Right member (or calculated right coordinate) is less than or equal to the Left member (or calculated left coordinate) or if the Bottom member (or calculated bottom coordinate) is less than or equal to the Top member (or calculated top coordinate).

Here
https://github.com/dotnet/runtime/blob/fcfb4b8a62b0bc30d9ed83a08cf9b322908c3973/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L941-L955

this should probably be
```c#
            // Check for arithmetic underflows & overflows.
            int newRight = left + srWindow.Right - srWindow.Left;
            if (left < 0 || newRight > csbi.dwSize.X - 1 || newRight < 0)
                throw new ArgumentOutOfRangeException(nameof(left), left, SR.ArgumentOutOfRange_ConsoleWindowPos);
            int newBottom = top + srWindow.Bottom - srWindow.Top;
            if (top < 0 || newBottom > csbi.dwSize.Y - 1 || newBottom < 0)
                throw new ArgumentOutOfRangeException(nameof(top), top, SR.ArgumentOutOfRange_ConsoleWindowPos);

            // Preserve the size, but move the position.
            srWindow.Bottom -= (short)(newBottom);
            srWindow.Right = (short)(newRight);
            srWindow.Left = (short)left;
            srWindow.Top = (short)top;

            bool r = Interop.Kernel32.SetConsoleWindowInfo(OutputHandle, true, &srWindow);
```
In other words -- we were correctly checking newRight against the bounds but then instead of setting the value to newRight, we were effectively setting it to newRight - 1. Instead, set the value we test.

SetWindowSize should be checked also.
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L960

**Comment by danmosemsft** [comment]
@daxian-dbw any interest in offering a PR? If not maybe I can look into that later.

**Comment by daxian-dbw** [comment]
@danmosemsft Of course! Thanks for the pointers 👍, I will submit a PR next week.

**Comment by danmosemsft** [comment]
Cool thanks! You might have a quick scan of other instances in the file (eg in SetWindowSiz)

Not sure whether a unit test is possible

**Comment by danmosemsft** [comment]
Fixed by #32243
